src/AWSS3.jl

```diff
@@ -410,6 +410,43 @@ end
 
 s3_delete(a...; b...) = s3_delete(global_aws_config(), a...; b...)
 
+"""
+    s3_delete_all_versions([::AbstractAWSConfig], bucket, path; kwargs...)
+
+Deletes all versions of object `path` from `bucket`, forwarding all `kwargs` to the
+internal per-version [`s3_delete`](@ref) call.
+
+To only delete one specific version, see [`s3_delete`](@ref); to delete all versions
+EXCEPT the latest version, see [`s3_purge_versions`](@ref).
+
+# API Calls
+
+- [`DeleteObject`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObject.html)
+- [`ListObjectVersions`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectVersions.html)
+
+# Permissions
+
+- [`s3:DeleteObject`](https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazons3.html#amazons3-DeleteObject)
+- [`s3:ListBucketVersions`](https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazons3.html#amazons3-ListBucketVersions)
+"""
+function s3_delete_all_versions(aws::AbstractAWSConfig, bucket, path; kwargs...)
+    # Because list_versions returns ALL keys with the given _prefix_, we need to
+    # restrict the results to ones with the _exact same_ key.
+    versions = [x["VersionId"] for x in s3_list_versions(bucket, path) if x["Key"] == path]
+    for version in versions
```

Comment:
Note that you can avoid materializing an array of version identifiers by looping over the full set of versions and skipping ones that don't meet your criteria:
```suggestion
    for object in s3_list_versions(aws, bucket, path)
        object["Key"] == path || continue
        version = object["VersionId"]
```
Also `s3_list_versions` should take the same configuration object passed to the calling function to ensure the right permissions are used.